ambry-router/src/main/java/com.github.ambry.router/PutManager.java

```diff
@@ -211,12 +208,14 @@ void close() {
    * 2. By the {@link ChunkFiller} thread when it exits abnormally.
    */
   void completePendingOperations() {
-    Iterator<PutOperation> iter = putOperations.iterator();
-    while (iter.hasNext()) {
-      PutOperation op = iter.next();
-      iter.remove();
-      NonBlockingRouter.completeOperation(op.getFuture(), op.getCallback(), null,
-          new RouterException("Aborted operation because Router is closed", RouterErrorCode.RouterClosed));
+    for (PutOperation op : putOperations) {
+      // There is a rare scenario where the operation gets removed from this set and gets completed concurrently by
```

Comment:
How did you figure this out(duplicate completion). Some tests which does parallel puts ?